Refactor documentation (#800)
minio/directpv: docs/architecture.md

# DirectPV CSI driver architecture

DirectPV is implemented as per [CSI specification](https://github.com/container-storage-interface/spec/blob/master/spec.md). It comes with below components run as Pods in Kubernetes.
* `Controller`
* `Node server`

When DirectPV contains legacy volumes from `DirectCSI`, below additional components run as Pods
* `Legacy controller `
* `Legacy node server`

## Controller

> Controller runs as `Deployment` Pods named `controller` which is three replicas in any Kubernetes nodes not limiting to where node server is running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers

Author: `controller` pods may (may not) run on nodes where `node server` pods are running. We have control on where `node server` pods to run, but not to `controller` pods.

> Controller runs as `Deployment` Pods named `controller` which is three replicas in any Kubernetes nodes not limiting to where node server is running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers

Reviewer: ```suggestion
The Controller runs as `Deployment` Pods named `controller`, which are three replicas located in any Kubernetes nodes, potentially including a pod with a node server running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers:
```

> Controller runs as `Deployment` Pods named `controller` which is three replicas in any Kubernetes nodes not limiting to where node server is running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers

Author: Actually I trimmed to remove `node-server` reference here.
```
The Controller runs as `Deployment` Pods named `controller`, which are three replicas located in any Kubernetes nodes. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers:
```

Does it look good?

* `Expand volume` - Controller server expands `DirectPVVolume` CRD after reversing requested storage space in `DirectPVDrive` CRD.

Below is a workflow diagram
```
┌────────────┐                                               ┌────────────┐
│            │ Create Event ┌─────────────┐ CreateVolume API │            │   ┌────────────────────┐
│            │------------->│     CSI     │----------------->│            │-->│  DirectPVDrive CRD │
│ Persistent │ Delete Event │             │ DeleteVolume API │            │   └────────────────────┘
│   Volume   │------------->│ Provisioner │----------------->│ Controller │
│   Claim    │              └─────────────┘                  │   Server   │
│   (PVC)    │                                               │            │   ┌────────────────────┐
│            │ Update Event ┌─────────────┐ ExpandVolume API │            │-->│ DirectPVVolume CRD │
│            │------------->│ CSI Resizer │----------------->│            │   └────────────────────┘
│            │              └─────────────┘                  └────────────┘
└────────────┘
```

## Legacy controller

> Legacy controller runs as `Deployment` Pods named `legacy-controller` which is three replicas in any Kubernetes nodes not limiting to where node server is running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers

Reviewer: ```suggestion
Legacy controller runs as `Deployment` Pods named `legacy-controller` which is three replicas in any Kubernetes nodes, potentially including a pod with a node server running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers
```

> Legacy controller runs as `Deployment` Pods named `legacy-controller` which is three replicas in any Kubernetes nodes not limiting to where node server is running. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers

Author: Actually I trimmed to remove `node-server` reference here.
```
Legacy controller runs as `Deployment` Pods named `legacy-controller`, which are three replicas located in any Kubernetes nodes. In the three replicas, one instance is elected to serve requests. Each pod contains below running containers:
```

Does it look good?

* `CSI provisioner` - Bridges legacy volume creation and deletion requests from `Persistent Volume Claim` to CSI controller.
* `Controller` - Honors CSI requests to delete and expand volumes. Create volume request is prohibited i.e. this controller works only for legacy volumes previously created in `DirectCSI`.
* `CSI resizer` - Bridges legacy volume expansion requests from `Persistent Volume Claim` to CSI controller.

> ### Legacy controller server

Reviewer: Ah, I see.
What you have is fine, then.
